Handle probe dependencies and hard errors better

pelwell opened the pull request:
#6642 is a report of intermittent failures for the PIO driver to contact the firmware. These failures can be traced to the failure of PIO driver to distinguish hard errors and transient failures. The commits in this PR fix the PIO driver in this regard, but also improve the firmware driver to avoid pointless retries in the event that the firmware is not compatible.

pelwell committed:
misc: rp1-pio: Error out on incompatible firmware

If the RP1 firmware has reported an error then return that from the PIO
probe function, otherwise defer the probing.

Link: https://github.com/raspberrypi/linux/issues/6642

Signed-off-by: Phil Elwell <phil@raspberrypi.com>

pelwell committed:
firmware: rp1: Linger on firmware failure

To avoid pointless retries, let the probe function succeed if the
firmware interface is configured correctly but the firmware is
incompatible. The value of the private drvdata field holds the outcome.

Link: https://github.com/raspberrypi/linux/issues/6642

Signed-off-by: Phil Elwell <phil@raspberrypi.com>

popcornmix commented on drivers/firmware/rp1.c:
```diff
@@ -270,19 +270,18 @@ static int rp1_firmware_probe(struct platform_device *pdev)
+		platform_set_drvdata(pdev, fw);
+	} else {
+		kfree(fw);
+		platform_set_drvdata(pdev, ERR_PTR(-ENOENT));
```
We do dereference this in `rp1_firmware_remove` (without a check). 
I'm not sure if that can happen, but looks like a potential pitfall.

pelwell replied:
Testing the obvious fix has uncovered some deficiencies in the end-of-life code. Thank goodness for runtime overlays.

pelwell committed:
mailbox: rp1: Don't claim channels in of_xlate

The of_xlate method saves the calculated event mask in the con_priv
field. It also rejects subsequent attempt to use that channel because
the mask is non-zero, which causes a repeated instantiation of a client
driver to fail.

The of_xlate method is not meant to be a point of resource acquisition.
Leave the con_priv initialisation, but drop the test that it was
previously zero.

Signed-off-by: Phil Elwell <phil@raspberrypi.com>

pelwell commented:
With the latest updates, the rp1 firmware driver and pio driver can be repeatedly instantiated and removed with no crashes or obvious leaks.
